Display driver addition

APorumb opened the pull request:
Added display driver using ssd_1306 IC from Zedboard.

Supports basic functions like print string, clear screen, display ON/OFF, etc..

ssd_1306 IC must be connected to SPI PS

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/analogdevicesinc/no-OS?pullRequest=1077) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/analogdevicesinc/no-OS?pullRequest=1077) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/analogdevicesinc/no-OS?pullRequest=1077) it.</sub>

buha commented on drivers/display/oled/oled.c:
```diff
@@ -0,0 +1,320 @@
+	                   0x0000000000000000,
+	                   0x0000000000000000,
+	                   0x0000000000000000,
+		               0x0000000000000000,
```
looks like there is a spacing/tab issue here

buha commented on drivers/display/oled/oled.h:
```diff
@@ -0,0 +1,95 @@
+/* Prints char string at selected position. */
+int32_t display_print_string(struct ssd_1306_dev *device, char *msg, uint8_t row, uint8_t column);
+
+
```
lots of blank lines ?

buha commented on drivers/display/oled/oled.h:
```diff
@@ -0,0 +1,95 @@
+#define DISP_OFF 0U
+
+//define display size
+#define ROWS_NB        4U
```
this should be configurable in oled_dev structure (which by the way looks like it's missing) because this generic oled.c/.h should work with displays of any size, theoretically

buha commented on drivers/display/oled/oled.h:
```diff
@@ -0,0 +1,95 @@
+/******************************************************************************/
+
+/* Turns display on/off */
+int32_t display_on_off(struct ssd_1306_dev *device, uint8_t on_off);
```
ssd_1306_dev is likely not what you want here, you probably want an oled_dev here and put ssd_1306 specific stuff in oled_dev->extra

buha commented on drivers/display/oled/oled.h:
```diff
@@ -0,0 +1,95 @@
+int32_t display_on_off(struct ssd_1306_dev *device, uint8_t on_off);
+
+/* Moves cursor to desired position. */
+int32_t move_cursor(struct ssd_1306_dev *device, uint8_t row, uint8_t column);
```
also, these oled.h functions should have like a common prefix oled_*

buha commented on drivers/display/ssd_1306/ssd_1306.c:
```diff
@@ -0,0 +1,178 @@
+ *                  Example: -1 - Initialization failed;
+ *                            0 - Initialization succeeded.
+*******************************************************************************/
+int32_t ssd_1306_init(struct ssd_1306_dev **device,
```
ssd_1306_* functions should probably be implemented as platform ops (operations / function pointers) much like we do for spi and i2c already

buha commented:
also, make sure to run astyle (ci/astyle.sh on a linux machine) and doxygen and see why the drivers build fails

adrimbarean commented:
For commit messages make sure to keep the lines limited at 75 characters and try to always have a title and body, even if the body is short.

adrimbarean commented on drivers/display/oled/oled.c:
```diff
@@ -0,0 +1,320 @@
+	int32_t ret = SUCCESS;
+	uint8_t command;
+
+	ret |= gpio_set_value(device->dc_pin, DC_CMD);
```
Should the SPI transfer happen if this doesn't happen? Usually we should exit upon an error unless we want to continue for some specific reason.

adrimbarean commented on drivers/display/oled/oled.c:
```diff
@@ -0,0 +1,320 @@
+	command[0] = 0x21;
+	command[1] = column*8;
+	command[2] = COL_NB * 8 - 1U;
+	ret |= spi_write_and_read(device->spi_desc, &command, 3U);
```
Could use 'ARRAY_SIZE' from util.h instead of 3U here.

adrimbarean commented on drivers/display/oled/oled.c:
```diff
@@ -0,0 +1,320 @@
+ *
+ * @return SUCCESS in case of success, negative error code otherwise.
+*******************************************************************************/
+int32_t display_clear(struct ssd_1306_dev *device, uint8_t rows, uint8_t columns)
```
Will this clear from position (0,0) or from current cursor position?
Is there any benefit here in using the 'display_print_ascii' API for more control?

adrimbarean commented on drivers/display/ssd_1306/ssd_1306.c:
```diff
@@ -0,0 +1,178 @@
+	ret |= gpio_set_value(device->vbat_pin, VBAT_ON);
+	usleep(2000U);
+
+	command[0] = 0xAE;
```
Maybe we would benefit from a list of commands as defines or enum in the header file.

adrimbarean commented on projects/oled_demo/Makefile:
```diff
@@ -0,0 +1,11 @@
+# Uncomment to use the desired platform
+# PLATFORM = xilinx
+# PLATFORM = altera
+# PLATFORM = aducm3029
+# PLATFORM = stm32
```
If this doesn't work for anything besides xilinx, maybe we should only have that option.

adrimbarean commented:
No need to fix one commit in a later commit in a pull-request. You can amend and force push while writing a comment here with what changed since last update.

amiclaus commented on drivers/display/oled/oled.c:
```diff
@@ -0,0 +1,320 @@
+*******************************************************************************/
+int32_t display_on_off(struct ssd_1306_dev *device, uint8_t on_off)
+{
+	int32_t ret = SUCCESS;
```
If you'll use the approach stated by @adrimbarean below, I don't think `ret` needs to be initialized here, since it will always initialized with the return value of `gpio_set_value`.

amiclaus commented on drivers/display/oled/oled.c:
```diff
@@ -0,0 +1,320 @@
+	command = (on_off == DISP_ON) ? 0xAF : 0xAE;
+	ret |= spi_write_and_read(device->spi_desc, &command, 1U);
+
+	return ret;
```
you can directly `return spi_write_and_read(device->spi_desc, &command, 1U);`

amiclaus commented on drivers/display/oled/oled.c:
```diff
@@ -0,0 +1,320 @@
+*******************************************************************************/
+int32_t move_cursor(struct ssd_1306_dev *device, uint8_t row, uint8_t column)
+{
+	int32_t ret = SUCCESS;
```
ditto. No need to initialized `ret` here.

amiclaus commented on drivers/display/oled/oled.c:
```diff
@@ -0,0 +1,320 @@
+
+	ret |= gpio_set_value(device->dc_pin, DC_CMD);
+
+	command = (on_off == DISP_ON) ? 0xAF : 0xAE;
```
what does `0xAF`, `0xAE` mean? maybe some suggestive macros should be added. This applies to most of the raw values used inside the driver.

amiclaus commented on drivers/display/oled/oled.c:
```diff
@@ -0,0 +1,320 @@
+
+	ret |= gpio_set_value(device->dc_pin, DC_DATA);
+
+	for(i=0; i<8; i++) //LSB at tx_char[8]
```
Is this comment left here on purpose?

buha replied:
Also, whitespaces and braces formatting is wrong, go through the linux style guideline first:
https://www.kernel.org/doc/html/v4.10/process/coding-style.html

amiclaus commented on drivers/display/oled/oled.c:
```diff
@@ -0,0 +1,320 @@
+
+	len = strlen(msg);
+	for(i=0; i<len; i++){
+		if(r<ROWS_NB){
```
nitpick: spacings are required here:
`if(r<ROWS_NB){` -> `if (r < ROWS_NB) {`

Check the entire code for similar issues.

amiclaus commented on drivers/display/oled/oled.c:
```diff
@@ -0,0 +1,320 @@
+	return ret;
+}
+
+
```
nit: extra line

amiclaus commented on drivers/display/oled/oled.c:
```diff
@@ -0,0 +1,320 @@
+	for(i=0; i<rows; i++)
+		for(j=0; j<columns; j++)
+			ret |= display_print_ascii(device, ' ', i, j);
+
```
nit: extra line

amiclaus commented on drivers/display/ssd_1306/ssd_1306.c:
```diff
@@ -0,0 +1,178 @@
+
+	dev = (struct ssd_1306_dev *)malloc(sizeof(*dev));
+	if (!dev)
+		return -1;
```
`return FAILURE` ?

amiclaus commented on drivers/display/ssd_1306/ssd_1306.c:
```diff
@@ -0,0 +1,178 @@
+ * 		               parameters.
+ *
+ * @return status - Initialization status.
+ *                  Example: -1 - Initialization failed;
```
`-1` -> FAILURE
`0` -> SUCCESS

amiclaus commented on projects/oled_demo/src.mk:
```diff
@@ -0,0 +1,18 @@
+        $(INCLUDE)/error.h                                              \
+        $(INCLUDE)/delay.h                                              \
+        $(DRIVERS)/platform/xilinx/gpio_extra.h                         \
+	$(DRIVERS)/platform/xilinx/spi_extra.h
```
nit: extra line needed

amiclaus commented on projects/oled_demo/src/app/main.c:
```diff
@@ -0,0 +1,135 @@
+	if (ret != SUCCESS)
+			return FAILURE;
+
+return 0;
```
return SUCCESS

amiclaus reviewed (commented):
Some comments on my side. Also check for the CI builds to track down detected issues for code style/project build/driver build etc.

buha commented on drivers/display/display.c:
```diff
@@ -0,0 +1,217 @@
+	dev = (struct display_dev *)malloc(sizeof(*dev));
+	if (!dev)
+		return FAILURE;
+	ret = gpio_get(&dev->reset_pin, param->reset_pin);
```
as discussed, these gpio_get calls should go to ssd_1306 init, they are specific to that chip only, they should not be in generic display.c which should only have traits general to all displays (like cols_nb, rows_nb)

buha commented on drivers/display/display.c:
```diff
@@ -0,0 +1,217 @@
+int32_t display_print_ascii(struct display_dev *device, uint8_t ascii,
+			    uint8_t row, uint8_t column)
+{
+
```
same here

buha commented on drivers/display/display.c:
```diff
@@ -0,0 +1,217 @@
+
+	for(i = 0; i < device->rows_nb; i++)
+		for(j = 0; j < device->cols_nb; j++)
+			ret |= display_print_ascii(device, ' ', i, j);
```
even if you've probably seen ret |= ...; in no-OS code, this is not ok - that's old code
 quit if there is an error

buha commented on drivers/display/display.c:
```diff
@@ -0,0 +1,217 @@
+	for(i = 0; i < len; i++) {
+		if(r < device->rows_nb) {
+			if(c < device->cols_nb) {
+				ret |= display_print_ascii(device, msg[i], r, c);
```
same here

buha commented on drivers/display/display.c:
```diff
@@ -0,0 +1,217 @@
+			} else {
+				c=0U;
+				r++;
+				ret |= display_print_ascii(device, msg[i], r, c);
```
same here

buha commented on drivers/display/display.h:
```diff
@@ -54,42 +54,80 @@
 #define DISP_ON  1U
```
these could be dropped and just use true/false

adrimbarean commented on drivers/display/display.c:
```diff
@@ -0,0 +1,187 @@
+
+	dev = (struct display_dev *)malloc(sizeof(*dev));
+	if (!dev)
+		return FAILURE;
```
If error here, we should `free(dev)`.

adrimbarean commented on drivers/display/display.c:
```diff
@@ -0,0 +1,187 @@
+		return FAILURE;
+	ret = spi_init(&dev->spi_desc, param->spi_ip);
+	if (ret != SUCCESS)
+		return FAILURE;
```
If error here, we should `free(dev)`.

adrimbarean commented on drivers/display/display.c:
```diff
@@ -0,0 +1,187 @@
+	dev->controller_ops = param->controller_ops;
+	dev->extra = param->extra;
+
+	ret = dev->controller_ops->init(dev);
```
`free(dev)` here too, if error.

amiclaus commented on drivers/display/ssd_1306/ssd_1306.c:
```diff
@@ -0,0 +1,383 @@
+		return FAILURE;
+	command[0] = 0x20;	// memory addressing mode
+	command[1] = 0x00;	// horizontal addressing
+	ret = spi_write_and_read(device->spi_desc, command, 2U);
```
`return spi_write_and_read(device->spi_desc, command, 2U);`

amiclaus commented on drivers/display/ssd_1306/ssd_1306.c:
```diff
@@ -0,0 +1,383 @@
+	if (ret != SUCCESS)
+		return FAILURE;
+	command = (on_off == true) ? DISP_ON : DISP_OFF;
+	ret = spi_write_and_read(device->spi_desc, &command, 1U);
```
ditto. return directly

amiclaus commented on drivers/display/ssd_1306/ssd_1306.c:
```diff
@@ -0,0 +1,383 @@
+	command[0] = 0x22;
+	command[1] = row;
+	command[2] = device->rows_nb - 1U;
+	ret = spi_write_and_read(device->spi_desc, &command, 3U);
```
ditto. return directly

amiclaus commented on drivers/display/ssd_1306/ssd_1306.c:
```diff
@@ -0,0 +1,383 @@
+	for(i = 0; i < 8; i++) {
+		ch[7-i] = (chr_map[ascii] >> (8*i)) & 0xFF;
+	}
+	ret = spi_write_and_read(device->spi_desc, ch, 8U);
```
ditto. return directly

amiclaus commented on projects/display_demo/src/app/main.c:
```diff
@@ -0,0 +1,157 @@
+	if (ret != SUCCESS)
+		return FAILURE;
+
+	ret = display_print_string(dev, msg, 0, 0);
```
ditto. return directly

adrimbarean commented on drivers/display/display.h:
```diff
@@ -0,0 +1,133 @@
+ * @struct display_dev
+ * @brief Display Device Descriptor.
+ */
+typedef struct display_dev {
```
I think we are not using `typedef` anymore, as per [Linux kernel coding style](https://www.kernel.org/doc/html/v4.10/process/coding-style.html#typedefs).

adrimbarean reviewed (approved):
Looks good to me. Just a few comments.

APorumb commented:
V5:

-added free(dev) after each failed init sequence
-at remove sequence, now removing first ssd_1306 dev and only after that remove spi
-now returning directly function's return
-eliminated typedefs

amiclaus commented on drivers/display/display.c:
```diff
@@ -0,0 +1,193 @@
+		return FAILURE;
+	ret = spi_init(&dev->spi_desc, param->spi_ip);
+	if (ret != SUCCESS) {
+		free(dev);
```
For future implementations, when you have multiple lines that repeat like these ones:
```
free(dev);
return FAILURE;
```
A conditional jump is preferred (at the end of the function).
In this function you have only 2 repetitions, so it should be fine as it is.

amiclaus reviewed (approved):
LGTM.

buha commented on drivers/display/display.c:
```diff
@@ -0,0 +1,193 @@
+ * @brief Turns display on/off.
+ *
+ * @param device - The device structure.
+ * @param on_off - TRUE = on
```
Where does the capital `TRUE` and `FALSE` are coming from ? which header ? I think the user should not care.

To simplify the user API, perhaps it would be better to have 2 functions in display.h/.c:
`display_on(dev)`
`display_off(dev)`

These in turn will call the corresponding op with true/false from <stdbool.h>

`display_on_off(dev, true);`
`display_on_off(dev, false);`

buha commented on drivers/display/ssd_1306/ssd_1306.c:
```diff
@@ -0,0 +1,371 @@
+};
+
+/* ASCII to bitmap */
+uint64_t chr_map[128] = {0x0000000000000000,
```
You could optimise this by splitting it into a chr_map[128][8]

I think optimizing it would make sense because the shifting and temporary storing at line 350 does not happen at init but happens in a function that is used at runtime.

buha commented on drivers/display/ssd_1306/ssd_1306.c:
```diff
@@ -0,0 +1,371 @@
+	ret = gpio_direction_output(extra->reset_pin, RST_OFF);
+	if (ret != SUCCESS)
+		return FAILURE;
+	usleep(2000U);
```
as a general rule, always document these delays by adding a comment of why is it needed, something like:
`Post reset delay, Treset=2ms`
where Treset is documented in datasheet

dbogdan commented on drivers/display/display.h:
```diff
@@ -0,0 +1,133 @@
+/** Prints char string at selected position. */
+int32_t display_print_string(struct display_dev *device, char *msg,
+			     uint8_t row, uint8_t column);
+
```
Let's implement display_print_char() too.

dbogdan commented on drivers/display/display.h:
```diff
@@ -0,0 +1,133 @@
+	/** Move cursor */
+	int32_t (*move_cursor)(struct display_dev *, uint8_t, uint8_t);
+	/** Print character by ascii number */
+	int32_t (*display_print_by_ascii)(struct display_dev *, uint8_t, uint8_t,
```
display_print_by_ascii -> display_print_char maybe

dbogdan commented on drivers/display/ssd_1306/ssd_1306.c:
```diff
@@ -0,0 +1,371 @@
+			 0x0000007f00000000,
+			 0x0000004136080000,
+			 0x0018080810101800,
+			 0xAA55AA55AA55AA55
```
Maybe move this to font.c?

dbogdan commented on drivers/display/ssd_1306/ssd_1306.h:
```diff
@@ -0,0 +1,110 @@
+#define DC_CMD     0U
+
+#define DISP_ON    0xAFU
+#define DISP_OFF   0xAEU
```
Move these MACROs to .c.

dbogdan commented on drivers/display/ssd_1306/ssd_1306.h:
```diff
@@ -0,0 +1,110 @@
+#include <stdbool.h>
+#include "display.h"
+#include "gpio.h"
+#include "spi.h"
```
Why do we need spi.h?

dbogdan commented on drivers/display/display.h:
```diff
@@ -0,0 +1,133 @@
+ */
+struct display_init_param {
+	/* SPI */
+	struct spi_init_param	   *spi_ip;
```
Why not having spi_ip in ssd_1306 init_param/desc?

APorumb commented:
V5:

-moved spi desc from device.c to ssd_1306.c
-added delay documentation
-added font.c used by ssd_1306.c
-added print_char function in display API
-refactored some ssd_1306 specific defines
-other minor changes

buha commented on util/font_8x8.c:
```diff
@@ -0,0 +1,175 @@
+	{0x00, 0x00, 0x22, 0x14, 0x14, 0x08, 0x00, 0x00},
+	{0x00, 0x02, 0x01, 0x59, 0x05, 0x02, 0x00, 0x00},
+	{0x3e, 0x41, 0x5d, 0x55, 0x4d, 0x51, 0x2e, 0x00},
+	{0x40, 0x7c, 0x4a, 0x09, 0x4a, 0x7c, 0x40, 0x00},//A
```
the comments are still here, let's either comment all or none, I'd opt for removing them

buha commented on drivers/display/display.c:
```diff
@@ -0,0 +1,209 @@
+{
+	struct display_dev *dev;
+	uint32_t ret;
+
```
Let's do error checking too, please #include "stderr.h" and modify all functions after local variables to have something like:
```
if (!device || !param)
    return -EINVAL;
```

Reason is, the common driver API from no-OS is used by users directly and we should ensure their program doesn't crash for silly reasons, like passing NULL pointers.

buha reviewed (commented):
looks good, some minor changes that I'd like to see implemented though

APorumb commented:
V6:

-removed comments from font_8x8.c
-added error checks for NULL pointers in display.c
